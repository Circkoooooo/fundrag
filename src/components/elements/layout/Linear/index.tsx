import { ElementComponentType } from '../../type'

const Linear: ElementComponentType = () => {
	return <div>linear</div>
}

Linear.componentName = '线性布局'

export default Linear
