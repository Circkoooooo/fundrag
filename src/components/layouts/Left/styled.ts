import styled from 'styled-components'

const WIDTH = ['100px', '400px']

interface LeftPanelProps {
	collapse: boolean
}

/**
 * 左面板容器
 */
export const LeftPanel = styled.div<LeftPanelProps>`
	width: ${(props) => (props.collapse ? WIDTH[0] : WIDTH[1])};
	height: 100%;
	background-color: #ffffff;
	transition: all ease-in-out 0.15s;
	box-shadow: 10px 0 20px rgba(0, 0, 0, 0.08);
	z-index: 1;
`
/**
 * 顶部区域
 */
export const HeaderMenu = styled.div`
	box-sizing: content-box;
	width: 100%;
	height: 40px;
	border-bottom: 2px solid rgba(0, 0, 0, 0.05);
`

/**
 * 折叠按钮
 */
export const CollapseButton = styled.div`
	width: 40px;
	height: 100%;
	margin-left: auto;
	cursor: pointer;
	user-select: none;
	padding: 8px;
`
