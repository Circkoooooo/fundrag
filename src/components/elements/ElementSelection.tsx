import styled from 'styled-components'
import elementBucket from './index'
import { ElementComponentType } from './type'

const ElementSelectionContainer = styled.div`
	width: 100%;
`
const Selection = styled.div`
	padding: 10px;
`

const SelectionName = styled.div`
	font-weight: bold;
	margin-bottom: 10px;
`

const SelectionMain = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 10px;
`

const ElementDisplayAround = styled.div`
	padding: 10px;
	background-color: #f0f0f0;
	border-radius: 4px;
	cursor: pointer;
	transition: all ease-in-out 0.15s;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`

interface ElementSelectionProps {
	pickElement?(element: ElementComponentType): void
}

/**
 * @description 将所有可以选择的组件渲染出来
 * @author onecirckoooooo
 * @date 2023/01/26 15:11
 */
const ElementSelection: React.FC<ElementSelectionProps> = ({ pickElement }) => {
	const renderGroupElements = () => {
		return elementBucket
			.filter((item) => item.elements.length !== 0)
			.map((item) => {
				return (
					<Selection key={crypto.randomUUID()}>
						<SelectionName>{item.name}</SelectionName>
						<SelectionMain>
							{item.elements.map((Element) => {
								return (
									<ElementDisplayAround
										key={crypto.randomUUID()}
										onClick={() =>
											pickElement && pickElement(Element)
										}
									>
										{Element.componentName}
									</ElementDisplayAround>
								)
							})}
						</SelectionMain>
					</Selection>
				)
			})
	}

	return (
		<ElementSelectionContainer>
			{renderGroupElements()}
		</ElementSelectionContainer>
	)
}

export default ElementSelection
