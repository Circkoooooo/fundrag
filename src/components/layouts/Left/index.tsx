import React, { useState } from 'react'
import {
	BsChevronDoubleRight as MenuCollapseRight,
	BsChevronDoubleLeft as MenuCollapseLeft,
} from 'react-icons/bs'
import { CollapseButton, HeaderMenu, LeftPanel } from './styled'

interface LeftProps {}

const Left: React.FC<LeftProps> = () => {
	const [isCollapse, setIsCollapse] = useState(true)

	return (
		<LeftPanel collapse={isCollapse}>
			<HeaderMenu>
				<CollapseButton onClick={() => setIsCollapse(!isCollapse)}>
					{isCollapse ? (
						<MenuCollapseRight
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					) : (
						<MenuCollapseLeft
							style={{
								width: '100%',
								height: '100%',
							}}
						/>
					)}
				</CollapseButton>
			</HeaderMenu>
		</LeftPanel>
	)
}

export default Left
