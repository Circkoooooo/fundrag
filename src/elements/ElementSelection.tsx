import { useEffect, useState } from 'react'
import styled from 'styled-components'
import elementBucket from './index'
import { chunkOutput, ElementFunctionComponentType } from './type'

const ElementSelectionContainer = styled.div`
	width: 100%;
`

const Selection = styled.div`
	padding: 10px;
`

const SelectionName = styled.div`
	font-weight: bold;
	margin-bottom: 10px;
`

const SelectionMain = styled.div`
	display: grid;
	grid-template-columns: 1fr 1fr;
	grid-gap: 10px;
`

interface ElementSelectionProps {
	pickElement?(element: ElementFunctionComponentType): void
}

const ElementDisplayAround = styled.div`
	padding: 10px;
	background-color: #f0f0f0;
	border-radius: 4px;
	cursor: pointer;
	transition: all ease-in-out 0.15s;
	white-space: nowrap;
	text-overflow: ellipsis;
	overflow: hidden;
`

/**
 * @description 将所有可以选择的组件渲染出来
 * @author onecirckoooooo
 * @date 2023/01/26 15:11
 */
const ElementSelection: React.FC<ElementSelectionProps> = ({ pickElement }) => {
	const [needRenderElements, setNeedRenderElements] = useState<chunkOutput[]>(
		[]
	)

	useEffect(() => {
		elementBucket.forEach((bucketItem) => {
			bucketItem.key = crypto.randomUUID()
		})
		setNeedRenderElements(
			elementBucket
				.filter((item) => item.elements.length !== 0) //筛选出含有子元素的容器元素
				.map((item) => {
					//给每一项element绑定unikey
					item.elements.forEach((element) => {
						element.defaultAppendProps.elementKey =
							crypto.randomUUID()
					})
					return item
				})
		)
	}, [])

	const renderGroupElements = () => {
		return needRenderElements.map((item) => {
			return (
				<Selection key={item.key}>
					<SelectionName>{item.name}</SelectionName>
					<SelectionMain>
						{item.elements.map((Element) => {
							return (
								<ElementDisplayAround
									key={Element.defaultAppendProps.elementKey}
									onClick={() =>
										pickElement && pickElement(Element)
									}
								>
									{Element.defaultAppendProps.componentName}
								</ElementDisplayAround>
							)
						})}
					</SelectionMain>
				</Selection>
			)
		})
	}

	return (
		<ElementSelectionContainer>
			{renderGroupElements()}
		</ElementSelectionContainer>
	)
}

export default ElementSelection
