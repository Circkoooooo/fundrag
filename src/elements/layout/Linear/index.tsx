import styled from 'styled-components'
import { useRef, useState } from 'react'
import { buildBaseProps } from '../..'
import { DefaultContainerProps, ElementFunctionComponentType, LayoutBaseStyleProperties } from '../../type'
import { LayoutBaseProps } from '../types'

const LinearContainer = styled.div<LayoutBaseStyleProperties>`
	${(props) => {
		return {
			width: props.width,
			height: props.height,
			backgroundColor: props.backgroundColor,
		}
	}}
`

type LinearProps = LayoutBaseProps & {}

const Linear: ElementFunctionComponentType<DefaultContainerProps, LinearProps> = ({ children, onClickEvent, defaultStyleProperties }) => {
	const styleProperties = useRef(defaultStyleProperties)

	return (
		<LinearContainer
			{...styleProperties.current}
			onClick={() => {
				onClickEvent &&
					onClickEvent({
						type: 'container',
						styleProperties: styleProperties.current,
					})
			}}
		>
			{children}
		</LinearContainer>
	)
}

Linear.defaultAppendProps = buildBaseProps('container', '线性组件')

export default Linear
