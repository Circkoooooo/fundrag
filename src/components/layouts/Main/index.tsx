import React from 'react'
import { ElementComponentType } from '../../elements/type'
import { MainContainer, MainScreen } from './styled'

interface MainProps {
	renderedElements?: ElementComponentType[]
}
const Main: React.FC<MainProps> = ({ renderedElements }) => {
	return (
		<MainContainer>
			<MainScreen>
				{renderedElements?.map((RenderedElement) => {
					return <RenderedElement key={crypto.randomUUID()} />
				})}
			</MainScreen>
		</MainContainer>
	)
}

Main.defaultProps = {
	renderedElements: [],
}

export default Main
