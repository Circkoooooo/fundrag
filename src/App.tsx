import { useState } from 'react'
import styled from 'styled-components'
import RenderElement from './components/RenderElement'
import { RenderedElementsType } from './components/RenderElement/types'
import ElementSelection from './elements/ElementSelection'
import { ClickEventAttributes, ClickEventElementType } from './elements/layout/types'
import { DefaultAppendProps, DefaultContainerProps, ElementFunctionComponentType } from './elements/type'
import Left from './layouts/Left'
import Main from './layouts/Main'
import Right, { ItemAttributes, ItemUnit } from './layouts/Right'

const Background = styled.div`
	width: 100%;
	height: 100%;
	background-color: #f0f0f0;
	display: flex;
`

export type SelectElement = {
	type: keyof ClickEventElementType
	key: string
}

function App() {
	const [appendElements, setAppendElements] = useState<RenderedElementsType<DefaultAppendProps, 'inline'>[]>([])
	const [layoutElements, setLayoutElements] = useState<RenderedElementsType<DefaultContainerProps, 'container'>[]>([])

	const [selectElement, setSelectElement] = useState<SelectElement | null>(null) //当前点击选中的元素
	const [itemAttributes, setItemAttributes] = useState<ItemAttributes[]>([] as ItemAttributes[]) //选中元素所有可修改属性的数组

	const renderElement = (element: ElementFunctionComponentType<DefaultAppendProps | DefaultContainerProps>) => {
		const elementType = element.defaultAppendProps.componentType

		if (elementType === 'inline' && selectElement?.type === 'layout' && selectElement.key !== undefined) {
			const appendElement: RenderedElementsType<DefaultAppendProps, 'inline'> = {
				key: crypto.randomUUID(),
				Element: element,
				containerKey: selectElement?.key, //寻找一个容器的key
				styleProperties: element.defaultAppendProps.defaultStyleProperties,
			}
			setAppendElements([...appendElements, appendElement])
		} else if (elementType === 'container') {
			console.log(element.defaultAppendProps)
			const layoutElement: RenderedElementsType<DefaultContainerProps, 'container'> = {
				key: crypto.randomUUID(),
				Element: element,
				styleProperties: element.defaultAppendProps.defaultStyleProperties,
			}
			setLayoutElements([...layoutElements, layoutElement])
		}
	}

	// 响应每次修改属性事件
	const onEditValue = (
		elementKey: string,
		type: 'container' | 'inline',
		attrObj: {
			itemTitle: string
			itemValue: string
			itemUnit: string
		},
		event: React.FormEvent<HTMLInputElement>,
		preItemAttributes?: ItemAttributes[]
	) => {
		const { itemTitle, itemUnit } = attrObj
		if (type === 'container') {
			const newLayoutElements = layoutElements.map((element) => {
				if (element.key === elementKey) {
					;(element.styleProperties as any)[itemTitle] = event.currentTarget.value + itemUnit
				}
				return element
			})

			if (preItemAttributes) {
				setItemAttributes(preItemAttributes)
			}
			setLayoutElements(newLayoutElements)
		}
	}

	/**
	 * 响应点击渲染过的元素后的事件
	 *
	 */
	const onPickElement = (elementKey: string, type: 'container' | 'inline', clickEventsAttribute: ClickEventAttributes) => {
		setSelectElement({
			type: clickEventsAttribute.type,
			key: elementKey,
		})

		if (clickEventsAttribute.styleProperties !== undefined) {
			const items: ItemAttributes[] = Object.entries(clickEventsAttribute.styleProperties).map((item, index) => {
				const pureValue = (item[1] as string).match(/\d+/g) //不附带属性的值
				const pureUnit = (item[1] as string).match(/\D+/g) //属性的单位

				const itemTitle = item[0]
				const itemValue = pureValue === null ? ((item[1] as string) === null ? '0' : (item[1] as string)) : pureValue[0]
				const itemUnit: ItemUnit = pureUnit === null ? 'px' : pureValue === null ? '' : (pureUnit[0] as ItemUnit)

				return {
					key: index.toString(),
					itemTitle,
					itemValue,
					itemUnit,
					editValue: (event: React.FormEvent<HTMLInputElement>, preItemAttributes?: ItemAttributes[]) => {
						onEditValue(
							elementKey,
							type,
							{
								itemTitle: item[0],
								itemValue,
								itemUnit,
							},
							event,
							preItemAttributes
						)
					},
				}
			})
			setItemAttributes(items)
		}
	}

	return (
		<Background>
			<Left>
				<ElementSelection pickElement={(element) => renderElement(element)} />
			</Left>
			<Main>
				<RenderElement
					{...{ layoutElements, appendElements }}
					onPickElement={(key: string, type: 'container' | 'inline', clickEventAttributes: ClickEventAttributes) => onPickElement(key, type, clickEventAttributes)}
				/>
			</Main>
			<Right {...{ itemAttributes, selectElement }} />
		</Background>
	)
}

export default App
