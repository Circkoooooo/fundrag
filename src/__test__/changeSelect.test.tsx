import renderer, { act } from 'react-test-renderer'
import App from '../App'
import Left from '../views/layouts/Left'
import ElementSelection from '../views/elements/ElementSelection'
import Main from '../views/layouts/Main'
import { MainScreen } from '../views/layouts/Main/styled'
import Linear from '../views/elements/layout/Linear/Linear'

describe('change select component after render', () => {
	const component = renderer.create(<App />)
	const instance = component.root
	const left = instance.findByType(Left)
	const elementSelection = left.findByType(ElementSelection)
	const main = instance.findByType(Main)

	it('cancel select', () => {
		let mainScreen = main.findByType(MainScreen)
		act(() => {
			elementSelection.props.onItemClick(Linear)
		})
		mainScreen = main.findByType(MainScreen)

		act(() => {
			mainScreen.props.onClick()
		})

		mainScreen = main.findByType(MainScreen)
		act(() => {
			elementSelection.props.onItemClick(Linear)
		})
		expect(mainScreen.props.children).toMatchSnapshot()
	})
})
