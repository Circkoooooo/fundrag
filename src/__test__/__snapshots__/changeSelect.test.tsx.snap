// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change select component after render cancel select 1`] = `
<React.Fragment>
  <component
    onSelectFn={[Function]}
  />
  <component
    onSelectFn={[Function]}
  />
</React.Fragment>
`;
